clients/src/main/java/org/apache/kafka/common/requests/OffsetFetchRequest.java

```diff
@@ -78,26 +85,107 @@ boolean isAllTopicPartitions() {
+                .setRequireStable(requireStable);
+            this.throwOnFetchStableOffsetsUnsupported = throwOnFetchStableOffsetsUnsupported;
+        }
+
         @Override
         public OffsetFetchRequest build(short version) {
             if (isAllTopicPartitions() && version < 2) {
                 throw new UnsupportedVersionException("The broker only supports OffsetFetchRequest " +
                     "v" + version + ", but we need v2 or newer to request all topic partitions.");
             }
-
+            if (data.groupIds().size() > 1 && version < 8) {
+                throw new NoBatchedOffsetFetchRequestException("Broker does not support"
+                    + " batching groups for fetch offset request on version " + version);
+            }
             if (data.requireStable() && version < 7) {
                 if (throwOnFetchStableOffsetsUnsupported) {
                     throw new UnsupportedVersionException("Broker unexpectedly " +
                         "doesn't support requireStable flag on version " + version);
                 } else {
                     log.trace("Fallback the requireStable flag to false as broker " +
-                                  "only supports OffsetFetchRequest version {}. Need " +
-                                  "v7 or newer to enable this feature", version);
+                        "only supports OffsetFetchRequest version {}. Need " +
+                        "v7 or newer to enable this feature", version);
 
                     return new OffsetFetchRequest(data.setRequireStable(false), version);
                 }
             }
-
+            if (version < 8) {
+                OffsetFetchRequestData oldDataFormat = null;
+                if (!data.groupIds().isEmpty()) {
+                    OffsetFetchRequestGroup group = data.groupIds().get(0);
+                    String groupName = group.groupId();
+                    List<OffsetFetchRequestTopics> topics = group.topics();
+                    List<OffsetFetchRequestTopic> oldFormatTopics = null;
+                    if (topics != null) {
+                        oldFormatTopics = topics
+                            .stream()
+                            .map(t ->
+                                new OffsetFetchRequestTopic()
+                                    .setName(t.name())
+                                    .setPartitionIndexes(t.partitionIndexes()))
+                            .collect(Collectors.toList());
+                    }
+                    oldDataFormat = new OffsetFetchRequestData()
+                        .setGroupId(groupName)
+                        .setTopics(oldFormatTopics)
+                        .setRequireStable(data.requireStable());
+                }
+                return new OffsetFetchRequest(oldDataFormat == null ? data : oldDataFormat, version);
+            }
+            // version 8 but have used old format of request, convert to version 8 of request
```

Comment:
Not sure I totally follow the suggestion here - The `version < 8` block followed by the subsequent `!data.groupIds().isEmpty()` if blog exists to convert the version to an old version if server side we are less than 8. Then we have the additional check for version 8 but using the old version of the request if the client sent the older version but the server supports the newer version, so we convert it to the newer version. What were you specifically thinking to change here to make the code more readable?

Reply:
At the moment, the code seems to do:
```
if (version < 8) {
 do-conversion-if-necessary
}
// version 8 but have used old format of request, convert to version 8 of request
do-conversion-if-necessary
```
My suggestion was to turn it into:
```
// convert data to use the appropriate version since version 8 uses different format
if (version < 8) {
 do-conversion-if-necessary
} else {
 do-conversion-if-necessary
}
```

Reply:
Ah okay, yes that makes sense, will make the change.